Tools/wasm/python.html

```diff
@@ -123,28 +124,44 @@
         this.xterm.open(container);
     }
 
-    handleReadComplete(lastChar) {
-        this.resolveInput(this.input + lastChar)
-        this.activeInput = false
-    }
-
     handleTermData = (data) => {
-        if (!this.activeInput) {
-            return
-        }
         const ord = data.charCodeAt(0);
-        let ofs;
+        data = data.replace(/\r|\n/g, "\r\n")  // Normalize line endings
 
+        // Handle pasted data
+        if (data.length > 1 && data.includes("\n")) {
+            let alreadyWrittenChars = 0;
+            // If line already had data on it, merge pasted data with it
+            if (this.input != '') {
+                this.inputBuffer.addData(this.input);
+                alreadyWrittenChars = this.input.length;
+                this.input = '';
+            }
+            this.inputBuffer.addData(data);
+            // If input is active, write the first line
+            if (this.activeInput) {
+                let line = this.inputBuffer.nextLine();
+                this.writeLine(line.slice(alreadyWrittenChars));
+                this.resolveInput(line);
+                this.activeInput = false;
+            }
+        // When input isn't active, add to line buffer
+        } else if (!this.activeInput) {
+            this.inputBuffer.addData(data);
         // TODO: Handle ANSI escape sequences
-        if (ord === 0x1b) {
+        } else if (ord === 0x1b) {
         // Handle special characters
         } else if (ord < 32 || ord === 0x7f) {
             switch (data) {
-                case "\r": // ENTER
+                case "\x0c": // CTRL+L
+                    this.clear();
+                    break;
+                case "\n": // ENTER
                 case "\x0a": // CTRL+J
                 case "\x0d": // CTRL+M
-                    this.xterm.write('\r\n');
-                    this.handleReadComplete('\n');
+                    this.resolveInput(this.input + this.writeLine('\n'));
```

Comment:
Could you explain why you are writing `\n` here if a) you normalize `\r` and `\n` to `\r\n` above, and b) you writeline so you write a `\r\n` after?

Reply:
maybe flush on '\n' 0xa only instead of '\r' (0xd) ? or better begin to handle canonical and raw terminal differently as they should.  LF should not be treated as EOL.( ncurses and ncursesw will be very soon available on wasm ).

Reply:
i also noticed that isatty() is returning 0 for  fd 0 / 1 / 2 and it should not
maybe try
```
    SYSCALLS.getStreamFromFD(0).tty = true;
    SYSCALLS.getStreamFromFD(1).tty = true;
    SYSCALLS.getStreamFromFD(2).tty = true;
```
at module init before trying to fix something that may not be broken :
```
>>> sys._emscripten_info
sys._emscripten_info(emscripten_version=(3, 1, 10), runtime='Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/100.0.4896.127 Safari/537.36', pthreads=False, shared_memory=False)
>>> 
>>> 
>>> print("                   123\r456",end="")
456>>>             123
>>> 
```

Reply:
> Could you explain why you are writing `\n` here if a) you normalize `\r` and `\n` to `\r\n` above, and b) you writeline so you write a `\r\n` after?

That inconsistency was unintentional. Python seems to have translated `\r\n` to simply `\n` in the strings and I hadn't noticed my mistake. In fact that `case` condition could never be reached due to `\r` always being converted to `\r\n` (due to the `data.length > 1` check above).

I just pushed a commit that normalizes lone `\r` to `\n` but keeps `\r\n` intact (and allows this `case` statement to actually run).

The `writeLine` doesn't actually write to Python. It writes to `xterm` only. The `xterm` session needs `\r\n` (it gets unhappy  otherwise: https://github.com/python/cpython/pull/91879#discussion_r857180078) but `writeLine` returns its input (`\n` in this case) which is then written to standard input. So `this.resolveInput` ends up being passed simply `this.input + '\n'`.